### PR TITLE
Unifying Header classes

### LINKED ISSUE #445: Unifying Header classes
We currently have 2 different classes for Headers: `RequestHeaders`, and `ResponseHeaders` that inherit from `HeadersBase` with our optimized dictionary implementation.

Unifying all into a `Headers` class would be preferred.

### PR DESCRIPTION
**Pull request by ealsur**
# Pull Request Template

## Description

We currently have 2 different classes for Headers: RequestHeaders, and ResponseHeaders that inherit from HeadersBase with our optimized dictionary implementation.

This PR unifies both classes into a single `Headers` class and unifies all tests.

## Closing issues

Closes #445

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+        internal string IsUpsert { get; set; }
+
+        [CosmosKnownHeaderAttribute(HeaderName = HttpConstants.HttpHeaders.OfferThroughput)]
+        internal string OfferThroughput { get; set; }
+
+        [CosmosKnownHeaderAttribute(HeaderName = HttpConstants.HttpHeaders.IfNoneMatch)]
+        internal string IfNoneMatch { get; set; }
+
+        [CosmosKnownHeaderAttribute(HeaderName = HttpConstants.HttpHeaders.PageSize)]
+        internal string PageSize { get; set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: >internal [](start = 8, length = 8)

Internals after public.

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+                this.ContentLengthAsLong = value != null ? long.Parse(value, CultureInfo.InvariantCulture) : 0;
+                this.contentLength = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets the Location for the current content in the <see cref="ResponseMessage"/>.
+        /// </summary>
+        [CosmosKnownHeaderAttribute(HeaderName = HttpConstants.HttpHeaders.Location)]
+        public virtual string Location { get; internal set; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: >public [](start = 8, length = 6)

Internal for now?

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: This was public, I just moved it, not sure what was the reason for it to be public though

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs (hunk 1 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: Done

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+            }
+            set
+            {
+                this.RequestCharge = string.IsNullOrEmpty(value) ? 0 : double.Parse(value, CultureInfo.InvariantCulture);
+            }
+        }
+
+        internal TimeSpan? RetryAfter;
+
+        internal SubStatusCodes SubStatusCode = SubStatusCodes.Unknown;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: I thought we are making SubStatusCode public?

File: Microsoft.Azure.Cosmos/src/Resource/CosmosException.cs [hunk 4]
```diff
@@ -113,7 +113,7 @@ public class CosmosException : Exception
         /// <summary>
         /// Gets the internal headers
         /// </summary>
-        internal virtual ResponseHeaders Headers { get; }
+        internal virtual Headers Headers { get; }
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosException.cs]: Having just Headers is confusing. 
```suggestion
        internal virtual Headers ResponseHeaders { get; }
```

File: Microsoft.Azure.Cosmos/src/Headers/Headers.cs (hunk 3 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: I wasn't aware of that, the request was to unify the classes, not change the accessors

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Headers/Headers.cs]: New PR

**Pull request merged**